command_cursor.c

```diff
@@ -105,23 +105,28 @@ int php_mongo_enforce_batch_size_on_command(zval *command, int size TSRMLS_DC)
 		return 0;
 	}
 
-	if (zend_hash_find(HASH_P(command), "cursor", 7, (void**) &zcursor) == FAILURE) {
-		zval *tmp_cursor;
-
-		MAKE_STD_ZVAL(tmp_cursor);
-		array_init(tmp_cursor);
-		zcursor = &tmp_cursor;
+	if (zend_hash_find(HASH_P(command), "cursor", 7, (void**) &tmp_cursor) == SUCCESS) {
+		zcursor = *tmp_cursor;
+	} else {
+		MAKE_STD_ZVAL(zcursor);
+		array_init(zcursor);
 
-		add_assoc_zval(command, "cursor", tmp_cursor);
+		add_assoc_zval(command, "cursor", zcursor);
 	}
 
-	if (Z_TYPE_PP(zcursor) != IS_ARRAY) {
-		php_mongo_cursor_throw(mongo_ce_CursorException, NULL, 32 TSRMLS_CC, "The cursor command's 'cursor' element is not an array");
+	if (Z_TYPE_P(zcursor) != IS_ARRAY && Z_TYPE_P(zcursor) != IS_OBJECT) {
+		php_mongo_cursor_throw(mongo_ce_CursorException, NULL, 32 TSRMLS_CC, "The cursor command's 'cursor' element is not an array or object");
 		return 0;
 	}
 
-	if (zend_hash_find(HASH_PP(zcursor), "batchSize", 10, (void**) &zbatchsize) == FAILURE) {
-		add_assoc_long(*zcursor, "batchSize", size);
+	if (!zend_hash_exists(HASH_P(zcursor), "batchSize", 10)) {
+		zval *tmp_batchsize;
+
+		MAKE_STD_ZVAL(tmp_batchsize);
```

Comment:
A note on the previous segfaults: `MAKE_STD_ZVAL(tmp_batchsize);` was actually doing garbage collection and allocating `tmp_batchsize` atop the original `zval` pointed to by the old `zval **zcursor` variable, if we previously allocated the command's cursor field with `MAKE_STD_ZVAL(zcursor);`.

A hacky solution would have been to save the `HashTable*` returned by `HASH_PP(zcursor)`, which would work because the cursor field is still there; however, avoiding the double pointers altogether seemed clearer.

Note that #805 would hide the segfault, since it removed calls to `php_mongo_enforce_batch_size_on_command()` on the collection and index enumeration functions; however, I added it back manually while testing just to confirm that the previous issue was resolved.